File: src/InorderSuccessorBST.java
class Solution {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    private TreeNode traverse(TreeNode node, TreeNode succ, int a){
        if(node == null) return succ;
        if(node.val > a && (succ == null || succ.val > node.val))  succ = node;
        TreeNode left = traverse(node.left, succ, a);
        TreeNode right = traverse(node.right, succ, a);
        if(left != null && right == null) return left;
        if(left == null && right != null) return right;
        if(left.val < right.val) return left;
        return right;
    }

    public TreeNode inorderSuccessor(TreeNode root, TreeNode p) {
        if(root == null) return null;
        return traverse(root, null, p.val);
    }
}

File: src/InorderSuccessorBST.py
class TreeNode(object):
     def __init__(self, x):
         self.val = x
         self.left = None
         self.right = None

class Solution(object):
    def inorderSuccessor(self, root, p):
        if not root:
            return None
        self.succ = None
        def traverse(node):
            if not node:
                return
            if node.val > p and (not self.succ or self.succ > node.val):
                self.succ = node.val
            traverse(node.left)
            traverse(node.right)

        traverse(root)
        return self.succ


sol = Solution()
root = TreeNode(2)
root.left = TreeNode(1)
print(sol.inorderSuccessor(root, 1))
